From 7e7752c30e7d413e42fc4c798144dcee8253848b Mon Sep 17 00:00:00 2001
From: Herbert Xu <herbert@gondor.apana.org.au>
Date: Mon, 1 Sep 2014 13:10:33 -0400
Subject: [crypto] drbg: simplify ordering of linked list in drbg_ctr_df

Message-id: <E1XORNR-00086W-Q9@gondolin.me.apana.org.au>
Patchwork-id: 90062
O-Subject: [PATCH 7/19] crypto: drbg - simplify ordering of linked list in drbg_ctr_df
Bugzilla: 1011562
RH-Acked-by: Jarod Wilson <jarod@redhat.com>
RH-Acked-by: Rafael Aquini <aquini@redhat.com>

RHEL7 Bugzilla #1011562

Upstream commit: c0eedf803467620fd3bdafccdd761990f6bd42e1

As reported by a static code analyzer, the code for the ordering of
the linked list can be simplified.

Reported-by: kbuild test robot <fengguang.wu@intel.com>
Signed-off-by: Stephan Mueller <smueller@chronox.de>
Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/crypto/drbg.c b/crypto/drbg.c
index faaa2ce..99fa8f8 100644
--- a/crypto/drbg.c
+++ b/crypto/drbg.c
@@ -516,13 +516,13 @@ static int drbg_ctr_df(struct drbg_state *drbg,
  S2.next = addtl;
 
  /*
-  * splice in addtl between S2 and S4 -- we place S4 at the end of the
-  * input data chain
+  * Splice in addtl between S2 and S4 -- we place S4 at the end
+  * of the input data chain. As this code is only triggered when
+  * addtl is not NULL, no NULL checks are necessary.
   */
  tempstr = addtl;
- for (; NULL != tempstr; tempstr = tempstr->next)
-  if (NULL == tempstr->next)
-   break;
+ while (tempstr->next)
+  tempstr = tempstr->next;
  tempstr->next = &S4;
 
  /* 10.4.2 step 9 */
-- 
1.7.1